common/geth/handle_error.go

```diff
@@ -0,0 +1,88 @@
+package geth
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/txpool"
+)
+
+type ChainConnFaults int64
+
+const (
+	SenderFault ChainConnFaults = iota
+	RPCFault
+	TooManyRequest
+	Ok
+	EVMFault
+)
+
+// this function accepts error message returned from API
+// decide if it is servers fault
+func HandleError(err error) ChainConnFaults {
+	if err == nil {
+		return Ok
+	}
+
+	// All geth core error codes from , https://pkg.go.dev/github.com/ethereum/go-ethereum@v1.13.14/core
+	// return false, as it is not RPC issue
+	if errors.Is(err, core.ErrNonceTooLow) ||
```

Comment:
Instead of matching errors directly, we should be matching the HTTP error. There are a lot more errors that can result from eth node than what's in `/core`. Also, does `strings.Contains(errMsg, "500")` actually match 500 error? What if the response returned 200 code with a JSON-RPC error code -32500 or JSON-RPC error message that happens to include `500`? 
Take a look at how [HTTP call](https://github.com/ethereum/go-ethereum/blob/master/rpc/client.go#L351) is made by the rpc client, and how the HTTP errors from [the call](https://github.com/ethereum/go-ethereum/blob/master/rpc/http.go#L173) are [returned](https://github.com/ethereum/go-ethereum/blob/master/rpc/http.go#L240). 
It looks like if HTTP returns something other than 2xx status, it will be wrapped by `HTTPError` with corresponding status code. I think it will be much easier to match this status code and attribute RPC fault if the status code isn't 2xx.

Reply:
My current thought on error handling is to differentiate the error into two cases. EVM error and non-EVM error. The non-EVM is a superset to connection error. For any connection error, it does not matter if the sender or the receiver is faulty, simply the connection does not work, and we should try a new one. 

Adding 2xx checks would be helpful to differentiate EVM-vs-NonEVM, it is just that I don't think there is a standard following by every party. For example, [alchemy](https://docs.alchemy.com/reference/error-reference) and [ankr](https://api-docs.ankr.com/reference/error-codes). Unless we have seen them, it is not clear what those error are. 

I did a parsing on rpc error from grafana, it looks like all EVM error (nonce) comes from core and txpool.

Reply:
Another weak reason to use EVM code is because optimism also does it. https://github.com/ethereum-optimism/optimism/blob/develop/op-service/txmgr/txmgr.go#L484

Reply:
Can you clarify what you mean by EVM error? 
An RPC provider could be totally functional and operational, but there could be a JSON-RPC error because the request isn't valid or depending on the network condition. We shouldn't attribute this to an RPC fault in this case. 
The list of error codes currently matched here only covers a fraction of the error cases that could be returned as a JSON-RPC error, and it doesn't seem reasonable to enumerate all of them (what happens if geth updates an error code or add another error?). 
Also, note that the failover usecase here is very different from `txmgr`. `txmgr` monitors/handles onchain transaction failures. It doesn't handle anything related to RPC provider failures.

Reply:
An EVM error is transaction related, for example the nonce is too low, price too low, etc. An EVM error implies the sender has made logic mistake somewhere, and as the result, no matter which RPC it sends, the result would be wrong. I think it is the exact case, we don't blame RPC. 

You raise another case that the content of transaction might be invalid. I agree it is another case we don't want to blame RPC, but I think our code is deterministic in the sense it does not construct it, unless we change ABI. 

The main problem this PR is trying to solve is to avoid RPC instability. For example, if network is unstable, regardless of which side, we should try to switch to another one. Since TCP does checksum, I don't think network instability can lead to corrupted data. 

Having addition RPC error code filtering would be nice, but I feel the EVM error checking is sufficient for now. But I am happy to change my mind. @jianoaix What do you think

Reply:
The main problem this PR is trying to solve should be to avoid RPC instability caused by the RPC provider, not the ethereum network or any other valid error. I think the cleanest way to distinguish the two faults is by using http status codes. (2xx responses can contain any json rpc errors, which means the fault is not at the RPC provider. 5xx responses mean the RPC provider is definitely at fault. 429 means RPC provider is not at fault but we should use a different provider)

Failover mechanism shouldn't distinguish whether it's an error coming from `core`/`txpool` package or any other package. What if we connect to a client which doesn't run geth / returns different error messages? 

To give a specific example of where this doesn't work is we can totally trigger `ErrExecutionReverted` or any other errors in [vm](https://github.com/ethereum/go-ethereum/blob/ba2dd9385c2a51134e520083dc732787a813b107/core/vm/errors.go#L31) package. In this case, cycling through multiple RPC providers would be pointless.

Reply:
The error handling is local to Geth made a valid good point. Although reth implemented identical error reaction as geth, see [here](https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/error.rs#L272C5-L272C16). But looking at other implementation like nethermind, it looks the error message is quite differently, see [here](https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs). Thank Ian for persistently pointing out my miss, it is good